### PR TITLE
Porting to LDAP v3

### PR DESCRIPTION
**Pull request by stefanmcshane**
Update to use ldapv3 per #13 

@johnweldon @vetinari Do you want to have a look over this as there were a few changes that were required.

### CODE REVIEW

File: go.sum911902081.tmp [hunk 1]
```diff
@@ -0,0 +1,2 @@
+github.com/go-asn1-ber/asn1-ber v1.3.1/go.mod h1:hEBeB/ic+5LoWskz+yKT7vGhhPYkProFKoKdwZRWMe0=
+github.com/go-ldap/ldap/v3 v3.1.7/go.mod h1:5Zun81jBTabRaI8lzN7E1JjyEl1g6zI6u9pd8luAK4Q=
```
Comment (johnweldon) [inline go.sum911902081.tmp]: Artefact of refactoring? File to be deleted?

File: ldif_test.go [hunk 2]
```diff
@@ -286,69 +287,84 @@ func TestParseModify(t *testing.T) {
...
 	}
 
-	if actual := attributes[1].Vals[0]; actual != "123 456 789 - 0" {
+	if actual := attributes[1].Modification.Vals[0]; actual != "123 456 789 - 0" {
 		t.Error("Expected", "123 456 789 - 0", "Actual", actual)
 	}
 
-	//ReplaceAttributes
-	attributes = l.Entries[0].Modify.ReplaceAttributes
+	// //ReplaceAttributes
```
Comment (johnweldon) [inline ldif_test.go]: double comment? (More below)

File: marshal.go [hunk 3]
```diff
@@ -70,49 +71,99 @@ func Marshal(l *LDIF) (data string, err error) {
...
+			// 	for _, v := range mod.Vals {
+			// 		ev, t := encodeValue(v)
+			// 		col := ": "
+			// 		if t {
+			// 			col = ":: "
+			// 		}
+			// 		data += foldLine(mod.Type+col+ev, fw) + "\n"
+			// 	}
+			// 	data += "-\n"
+			// }
```
Comment (johnweldon) [inline marshal.go]: Do we need to keep these commented lines?

File: go.mod [hunk 4]
```diff
@@ -0,0 +1,10 @@
+module github.com/go-ldap/ldif
+
+go 1.14
+
+require (
+	github.com/go-asn1-ber/asn1-ber v1.4.1 // indirect
+	github.com/go-ldap/ldap v3.0.3+incompatible
```
Comment (johnweldon) [inline go.mod]: Could this possibly be cleaned out - do we know why this line gets generated?

File: go.mod [hunk 5]
```diff
@@ -0,0 +1,10 @@
...
+	github.com/go-ldap/ldap/v3 v3.1.7
+	gopkg.in/asn1-ber.v1 v1.0.0-20181015200546-f715ec2f112d // indirect
```
Comment (johnweldon) [inline go.mod]: This line seems to be leftover from some transitive import - maybe it can be removed now?

File: go.sum [hunk 6]
```diff
@@ -0,0 +1,10 @@
+github.com/go-asn1-ber/asn1-ber v1.3.1 h1:gvPdv/Hr++TRFCl0UbPFHC54P9N9jgsRPnmnr419Uck=
+github.com/go-asn1-ber/asn1-ber v1.3.1/go.mod h1:hEBeB/ic+5LoWskz+yKT7vGhhPYkProFKoKdwZRWMe0=
+github.com/go-asn1-ber/asn1-ber v1.4.1 h1:qP/QDxOtmMoJVgXHCXNzDpA0+wkgYB2x5QoLMVOciyw=
+github.com/go-asn1-ber/asn1-ber v1.4.1/go.mod h1:hEBeB/ic+5LoWskz+yKT7vGhhPYkProFKoKdwZRWMe0=
+github.com/go-ldap/ldap v3.0.3+incompatible h1:HTeSZO8hWMS1Rgb2Ziku6b8a7qRIZZMHjsvuZyatzwk=
+github.com/go-ldap/ldap v3.0.3+incompatible/go.mod h1:qfd9rJvER9Q0/D/Sqn1DfHRoBp40uXYvFoEVrNEPqRc=
+github.com/go-ldap/ldap/v3 v3.1.7 h1:aHjuWTgZsnxjMgqzx0JHwNqz4jBYZTcNarbPFkW1Oww=
+github.com/go-ldap/ldap/v3 v3.1.7/go.mod h1:5Zun81jBTabRaI8lzN7E1JjyEl1g6zI6u9pd8luAK4Q=
+gopkg.in/asn1-ber.v1 v1.0.0-20181015200546-f715ec2f112d h1:TxyelI5cVkbREznMhfzycHdkp5cLA7DpE+GKjSslYhM=
+gopkg.in/asn1-ber.v1 v1.0.0-20181015200546-f715ec2f112d/go.mod h1:cuepJuh7vyXfUyUwEgHQXw849cJrilpS5NeIjOWESAw=
```
Comment (johnweldon) [inline go.sum]: Ditto

File: go.sum [hunk 7]
```diff
@@ -0,0 +1,10 @@
+github.com/go-asn1-ber/asn1-ber v1.3.1 h1:gvPdv/Hr++TRFCl0UbPFHC54P9N9jgsRPnmnr419Uck=
+github.com/go-asn1-ber/asn1-ber v1.3.1/go.mod h1:hEBeB/ic+5LoWskz+yKT7vGhhPYkProFKoKdwZRWMe0=
+github.com/go-asn1-ber/asn1-ber v1.4.1 h1:qP/QDxOtmMoJVgXHCXNzDpA0+wkgYB2x5QoLMVOciyw=
+github.com/go-asn1-ber/asn1-ber v1.4.1/go.mod h1:hEBeB/ic+5LoWskz+yKT7vGhhPYkProFKoKdwZRWMe0=
+github.com/go-ldap/ldap v3.0.3+incompatible h1:HTeSZO8hWMS1Rgb2Ziku6b8a7qRIZZMHjsvuZyatzwk=
+github.com/go-ldap/ldap v3.0.3+incompatible/go.mod h1:qfd9rJvER9Q0/D/Sqn1DfHRoBp40uXYvFoEVrNEPqRc=
```
Comment (johnweldon) [inline go.sum]: Ditto

**Comment by stefanmcshane** [comment]
@johnweldon All changes made. 
Travis jobs failing on golint. Any ideas?

File: marshal.go (hunk 3 above)
Comment (stefanmcshane) [reply marshal.go]: Deleted

File: go.mod (hunk 5 above)
Comment (stefanmcshane) [reply go.mod]: Go mod tidy fixed all this

File: go.mod (hunk 4 above)
Comment (stefanmcshane) [reply go.mod]: Generated when running travis jobs/make files manually

File: .travis.yml [hunk 8]
```diff
@@ -2,17 +2,20 @@ language: go
...
     allow_failures:
         - go: tip
 go_import_path: github.com/go-ldap/ldif
 install:
-    - go get gopkg.in/asn1-ber.v1
-    - go get gopkg.in/ldap.v2
+    - go get github.com/go-ldap/ldap/v3
     - go get code.google.com/p/go.tools/cmd/cover || go get golang.org/x/tools/cmd/cover
-    - go get github.com/golang/lint/golint || true
+    - go get golang.org/x/lint || true
```
Comment (johnweldon) [inline .travis.yml]: maybe use `go install golang.org/x/lint` instead here? Although go get seems to work fine locally.

Comment (stefanmcshane) [reply .travis.yml]: @johnweldon Got there in the end - new import path for go1.10 onwards golang.org/x/lint/golint is now enforced by the project

Comment (johnweldon) [reply .travis.yml]: Nice, thanks!
